### PR TITLE
Log.iterator Function

### LINKED ISSUE #231: Log Iterator Function
See: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Iterators_and_Generators
See: https://github.com/orbitdb/orbit-db-eventstore/blob/master/src/EventStore.js#L27

Entry / Heads / Tails index work??

### PR DESCRIPTION
**Pull request by aphelionz**
## Description

This PR creates the `Log.iterator`  function. In the end, the decision was made to simply utilize the `Log.traverse` inside the iterator and return a generator function, which is syntactic sugar for `Symbol.iterator`. This way, any improvements made to log traversal will bubble up to the iterator.

## Testing
For examples, look at `test/log-iterator.spec.js`

## TODO
- [ ] Documentation, particularly examples

## Next Steps

One way to improve this would be to make it async and utilize ipfs.dag within the generator function to make it lazy in its fetching, but that's not within the scope of this PR.

Resolves #231

### CODE REVIEW

File: src/log.js [hunk 1]
```diff
@@ -270,6 +270,30 @@ class Log extends GSet {
...
+   *
+   * @param {Object} options
+   * @param {string} options.gte Beginning hash of the iterator
+   * @param {string} options.lte Ending hash of the iterator
+   * @param {number} options.number Number of entried to return to / from the gte / lte hash
+   * @returns {Symbol.Iterator} Iterator object containing log entries
+   *
+   */
+  iterator ({ gte = undefined, lte = undefined, amount = -1 } = {}) {
+    if(!gte && !lte) throw LogError.GteOrLteNotDefinedError()
```
Comment (haadcode) [inline src/log.js]: We probably don't need to force the `gte` or `lte`. If no options given, default should be to start iterating from the heads? What do you think?

We also want to have `gt` and `lt` in addition to `gte/lte` (ie. "iterate up until <gt/lt>, don't include the given hash").

Comment (aphelionz) [reply src/log.js]: Sounds good to me!

Comment (aphelionz) [reply src/log.js]: @haadcode should we throw an error in any event at all? like if gt and gte (or lt and lte) are passed simultaneously?

**Comment by shamb0t** [comment]
@aphelionz scaffolding lgtm :+1:

File: src/log-errors.js [hunk 2]
```diff
@@ -4,10 +4,13 @@ const IPFSNotDefinedError = () => new Error('IPFS instance not defined')
 const LogNotDefinedError = () => new Error('Log instance not defined')
 const NotALogError = () => new Error('Given argument is not an instance of Log')
 const CannotJoinWithDifferentId = () => new Error('Can\'t join logs with different IDs')
+const GteOrLteNotDefinedError =
```
Comment (haadcode) [inline src/log-errors.js]: Is this used anywhere anymore?

File: src/log.js [hunk 3]
```diff
@@ -181,7 +181,7 @@ class Log extends GSet {
     return this._entryIndex[entry.cid || entry] !== undefined
   }
 
-  traverse (rootEntries, amount = -1) {
+  traverse (rootEntries, amount = -1, endHash) {
```
Comment (haadcode) [inline src/log.js]: Does the result include the `endHash` entry?

File: test/log-iterator.spec.js [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+    describe('Basic iterator functionality', () => {
+      let log1
+
+      beforeEach(async () => {
+        log1 = new Log(ipfs, testIdentity, { access: testACL, logId: 'X' })
+
+        for (let i = 0; i <= 100; i++) {
+          await log1.append('entry' + i)
+        }
+      })
```
Comment (haadcode) [inline test/log-iterator.spec.js]: Would be good to test with forking/joining log also, eg. with this https://github.com/orbitdb/ipfs-log/blob/master/test/utils/log-creator.js#L6. It may reveal some new design considerations and edge cases.

File: test/log-iterator.spec.js [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+          assert.strictEqual(entry.payload, 'entry' + (79 - i++))
+          count++
+        }
+        assert.strictEqual(count, amount)
+      })
+
+      /* eslint-disable camelcase */
+      it('iterates with combinations of lt, gt, lte, and gte', async () => {
+        // lt + gt
+        let it_lt_gt = log1.iterator({
```
Comment (haadcode) [inline test/log-iterator.spec.js]: Small stylistic remark: we use camelCase elsewhere in code. Also, perhaps the variables names here can be given more descriptive names?

File: test/log-iterator.spec.js [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+        let count = 0
+        for (let entry of it) {
+          assert.strictEqual(entry.payload, 'entry' + (79 - i++))
+          count++
+        }
+        assert.strictEqual(count, amount)
+      })
+
+      /* eslint-disable camelcase */
+      it('iterates with combinations of lt, gt, lte, and gte', async () => {
```
Comment (haadcode) [inline test/log-iterator.spec.js]: This test would benefit from being broken down to multiple `it`s. Perhaps one `it` per iterator call or similar.

File: test/log-iterator.spec.js [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+
+        // lt + gte
+        let it_lt_gte = log1.iterator({
+          gte: 'zdpuAyw2rPe5ML6YmYe7DEyAq3CbWH3QpWQ4P6F4ZFroEebzz',
+          lt: 'zdpuAxv8bUaFNmCp51zN3AwCLy83FfVCcp7GeTwCdtpVddziP'
+        })
+        let it_lt_gte_array = [...it_lt_gte]
+
+        // neither hash should appear in the array
+        assert.strictEqual(it_lt_gte_array.map(e => e.hash)
```
Comment (haadcode) [inline test/log-iterator.spec.js]: The mapping here and below could be done in one place before the asserts. Saves a few cycles :)

File: test/log-iterator.spec.js [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+        let it = log1.iterator({
+          lte: 'zdpuAx5FC3cgoQSW7oXpBw6x4YiuDuaF9BBgncguzvndP8tDL'
+        })
+
+        assert.strictEqual([...it].length, 68)
+
+        let i = 0
+        for (let entry of it) {
+          assert.strictEqual(entry.payload, 'entry' + (67 - i++))
+        }
```
Comment (haadcode) [inline test/log-iterator.spec.js]: The check for the right order as done in these loops could be separated to their own tests.

**Comment by haadcode** [comment]
Looking really good @aphelionz! 👍 🎉 Left some comments on how this PR could be improved a bit. Functionality-wise looks good to me, but would be good to test with the forking/joining log to see it's solid in branching log cases too.

File: test/log-iterator.spec.js (hunk 5 above)
Comment (aphelionz) [reply test/log-iterator.spec.js]: They are meant to be descriptive - it_lt_gte -> iterator with lt and gte. Camelcase didnt make sense and that's why I made the exception for this one test.

File: test/log-iterator.spec.js (hunk 6 above)
Comment (aphelionz) [reply test/log-iterator.spec.js]: I can do that, yeah... that would fix the variable naming problem above as well.

File: src/log.js (hunk 3 above)
Comment (aphelionz) [reply src/log.js]: Yes it does

File: test/log-iterator.spec.js (hunk 4 above)
Comment (aphelionz) [reply test/log-iterator.spec.js]: Added tests for forked/joins based on the logCreator work. Have a look and resolve this convo if you like it :)

Comment (haadcode) [reply test/log-iterator.spec.js]: Looks good now 👍 Thanks!

File: .circleci/config.yml [hunk 9]
```diff
@@ -39,9 +38,9 @@ jobs:
     # 'See docs on artifact collection here https://circleci.com/docs/2.0/artifacts/'
     - run: mkdir -p $CIRCLE_ARTIFACTS $CIRCLE_TEST_REPORTS
     # This is based on your 1.0 configuration file or project settings
-    - run:
-        working_directory: ~/orbitdb/ipfs-log
-        command: nvm install 10.15.0 && nvm alias default 10.15.0
+    #  - run:
+    #    working_directory: ~/orbitdb/ipfs-log
+    #    command: nvm install 10.15.0 && nvm alias default 10.15.0
```
Comment (haadcode) [inline .circleci/config.yml]: Was anything changed here? I can't spot the difference...

**Comment by haadcode** [comment]
Thanks @aphelionz, looks good to me now! 👍 One thing you might want to consider is to squash the commits to clean up the "fix: circleci" commits :)

Good to merge from my side, but would love to get one more look from @shamb0t too.

File: test/log-iterator.spec.js [hunk 10]
```diff
@@ -0,0 +1,344 @@
...
+  stopIpfs
+} = require('./utils')
+
+let ipfs, testIdentity, testIdentity2, testIdentity3, testIdentity4
+
+Object.keys(testAPIs).forEach((IPFS) => {
+  describe('Log - Iterator (' + IPFS + ')', function () {
+    this.timeout(config.timeout)
+
+    const testACL = new AccessController()
```
Comment (shamb0t) [inline test/log-iterator.spec.js]: `testACL` isn't really necessary anymore as it is the same as the Log default value

**Comment by shamb0t** [comment]
One small comment but otherwise lgtm :ok_hand:

File: test/log-iterator.spec.js (hunk 10 above)
Comment (aphelionz) [reply test/log-iterator.spec.js]: Removed!

File: .circleci/config.yml (hunk 9 above)
Comment (aphelionz) [reply .circleci/config.yml]: You can disregard

**Comment by aphelionz** [comment]
Merging this despite the circleci puppeteer issues
